### ISSUE TITLE
Helper function namespacing

### ISSUE THREAD

**Issue by zoul** [pointfreeco/swift-html#20]
First, thank you for the library, it’s a real joy to use! 🦄 The context-driven autocompletion feels like someone is reading my mind sometimes :) And the sheer amount of implemented HTML is impressive. (Did you use some kind of code generation or automation?)

The only thing I have trouble with is what all big DSLs in Swift seem to struggle with, and that’s namespace clashes. The compiler seems to get into a corner very often, ending up with hard-to-read error messages:

```swift
// error: expected expression in list of expressions
// error: consecutive statements on a line must be separated by ';'
// error: expected identifier in class declaration
let foo = small(class("sample"), "whatever")
```

Here the problem is easy to spot, but with more complex expressions one has to decompose them into pieces when the whole thing doesn’t compile at the first try. I usually end up prefixing the troublesome cases (`Html.class`), but it’s slightly annoying. People who are just starting with DSLs might be pretty confused by error messages such as “being unable to infer T somewhere”, there’s not much to hang on. Do you also run into this problem often? Is there some better solution than prefixing a lot?

**Comment by mbrandonw** [comment]
Hey @zoul, thanks for trying out the library!

Have you tried escaping with back ticks? You can do `\`class\`(“sample”)` and that usually helps the compiler.

**Comment by mbrandonw** [comment]
Hm github isn’t escaping that correctly. Try this:

    `class`(“sample”)

**Comment by mbrandonw** [comment]
Just wanted to check in... did this help at all @zoul?

**Comment by zoul** [comment]
Sorry, got sidetracked by life. Yes, that helps! Thank you. I was aware of you using the same trick in the framework, but somehow didn’t realize I can do the same.

Although I have to admit I‘m not completely happy with it. There’s a lot of punctuation in the source as it is, and escaping things with `` routinely doesn’t help. Do you think there could be some other tricks in your sleeve? Possibly even something that’s not in the language right now but could be added? (Apart from macros, that is.) How do other DSL-positive languages solve this? (Or should I simply add a `klass` helper and stop complicating things?)

Also, still on this topic, did you think about supporting single-item arguments? Given this definition:

```swift
public func h2(_ attribs: [Attribute<Tag.H2>], _ content: [Node]) -> Node {
  return element("h2", attribs, content)
}
```

I can only call the function like this:

```swift
h2([`class`("main-header")], [.text(article.title)])
```

Which is a lot of noise if I only have one class and one child node. It would be nice if I could write this instead:

```swift
h2(`class`("main-header"), .text(article.title))
```

In other words, use a single item where an array is expected. I understand it’t pure sugar, but sugar is an important part of DSLs, so 🤷‍♂️

**Comment by mbrandonw** [comment]
Yeah I definitely agree that there is a lot of punctuation, and anything we can do to reduce would be nice.

 I think `klass` is definitely an option, tho I never really liked that from Ruby. You could also define a `class_` helper, which is a pretty common style in Haskell and Purescript, which don't have function overloads. I wish Swift could try a little harder to figure out what is meant by `class("sample")`, and perhaps it should even be filed as a bug or a request for improvement. 

> In other words, use a single item where an array is expected. I understand it’t pure sugar, but sugar is an important part of DSLs, so 🤷‍♂️

Yeah that does read better, but a few problems I can see:

* It can be annoying to use the API without arrays, and then the moment you need multiple children/attributes you have to wrap everything in an array. It's a small annoyance, but after it happens a few dozen times you start to feel like the API is no longer pulling its weight. And I say this as someone who has run into this a _bunch_ with our [`View`](https://github.com/pointfreeco/swift-web/blob/442c64e209cfc675e72e37c36e9bc9f97888a2b8/Sources/View/View.swift) type, which is a function `(D) -> [Node]`. We have a convenience initializer that allows just returning a `Node` and it gets wrapped into an array. I'm constantly fighting that API, so much so that we plan to get rid of it and always use functions into `[Node]`.
* Adding an overload to a bunch of the HTML functions that take a single attribute or child element might increase the compile time. I don't know this for sure, but it seems reasonable. One thing we could do to find out is add a few of the overloads for some of the most common HTML tags, and then see if the code on https://www.github.com/pointfreeco/pointfreeco takes longer to compile.

**Comment by zoul** [comment]
> I wish Swift could try a little harder to figure out what is meant by class("sample"), and perhaps it should even be filed as a bug or a request for improvement.

I’ve tried to pitch this [on the Swift Forums](https://forums.swift.org/t/is-it-possible-to-disambiguate-class-better/18317), so let’s see if we can come up with something there.

> Adding an overload to a bunch of the HTML functions that take a single attribute or child element might increase the compile time.

I was also afraid that would be the case. And the increasing number of overloads often decreases the quality of error messages, since the compiler gets confused more easily.

I will close this and reopen if I can come up with concrete improvements. Thank you!
